src/stream/src/executor/exchange/input.rs

```diff
@@ -203,10 +205,29 @@ impl RemoteInput {
                                 barrier.mutation.is_none(),
                                 "Mutation should be erased in remote side"
                             );
-                            let mutation = local_barrier_manager
-                                .read_barrier_mutation(barrier)
+                            let mutation_subscriber =
+                                mutation_subscriber.get_or_insert_with(|| {
+                                    local_barrier_manager
+                                        .subscribe_barrier_mutation(self_actor_id, barrier)
+                                });
+
+                            let mutation = mutation_subscriber
+                                .recv()
                                 .await
-                                .context("Read barrier mutation error")?;
+                                .ok_or_else(|| {
+                                    anyhow!("failed to receive mutation of barrier {:?}", barrier)
+                                })
+                                .and_then(|(prev_epoch, mutation)| {
+                                    if prev_epoch != barrier.epoch.prev {
+                                        Err(anyhow!(
```

Comment:
It seems to be a bug if this happens. What about `unreachable!`?

Reply:
Or add assertion for it?